src/acc/opencl/smm/tune_multiply.py

```diff
@@ -214,6 +212,7 @@ def save_final_config(self, configuration):
             with open(ofilename, "w") as ofile:
                 json.dump(config, ofile)
                 ofile.write("\n")  # append newline at EOF
+                ofile.close()
```

Comment:
the context manager (which you get by using `with open(...)` will also try to close the file when leaving scope (but is usually able to cope with an already closed file). I'd rather reduce the indent level to trigger it:

```python
with open(ofilename, "w") as ofile:
    json.dump(config, ofile)
    ofile.write("\n")
print(f"Result achieving {self.gflops}")
```

with regard to string concatenation: I would try to avoid it as best as possible (since slow).
The cleanest solution if you can target Python 3.6 and later are f-strings: `f"content: {some_var}"`.
If you have to support Python 3.5 and earlier I'd rather use `"{}".format(some_var)` rather than string concatenation.

Reply:
> will also try to close the file when leaving scope

That I know, but I wanted to call `merge_into_csv` in underneath of the with-block. However, `merge_into_csv` reads that otherwise still opened file (which can/will fail).

Probably, I should call `merge_into_csv` after the with-block...

Reply:
I have replaced string concatenation with formatting a string.